### PR TITLE
(MODULES-4188) Add UUID generation function

### PR DESCRIPTION
**Pull request by petems**
* Generates UUID based on a given string and 
optional seed

### CODE REVIEW

File: lib/puppet/parser/functions/uuid_generate.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    Creates a UUID based on a given string and an optional seed value.
+
+    For example, to generate a UUID based on the FQDN of a system:
+
+    Usage:
+
+      $uuid = uuid_generate($::fqdn)
+
+    The seed is optional. If given it must be a string.
```
Comment (hunner) [inline lib/puppet/parser/functions/uuid_generate.rb]: It should mention if the seed should be numeric, hex, or an arbitrary string.

File: lib/puppet/parser/functions/uuid_generate.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    END
+
+    if args.length == 0
+      raise(ArgumentError, "uuid_generate: No arguments given")
+    elsif args.length == 1
+      seed = nil
+      fqdn = args[0]
+    elsif args.length == 2
+      fqdn = args[0]
```
Comment (hunner) [inline lib/puppet/parser/functions/uuid_generate.rb]: Technically it doesn't have to be an FQDN so it should be called `input_string` or something.

File: lib/puppet/parser/functions/uuid_generate.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      seed = args[1]
+      unless seed.is_a?(String)
+        raise(Puppet::ParseError, 'uuid_generate(): seed argument must be a string')
+      end
+    else
+      raise(ArgumentError, "uuid_generate: Too many arguments given (#{args.length})")
+    end
+
+    # The UUID for our namespace (6ba7b810-9dad-11d1-80b4-00c04fd430c8)
+    ns_uuid = "\x0b\x7a\x81\xff\xdb\x8d\x42\xfe\x8d\x9f\x76\x8e\xa5\xb8\xed\x1a"
```
Comment (hunner) [inline lib/puppet/parser/functions/uuid_generate.rb]: This UUID name space ID is only for prefixing FQDNs and as this is an "uuid_generate puppet function arbitrary string" UUID, we should create a new namespace specifically for this function (I imagine anything will do).

As per RFC-4122:
> Allocate a UUID to use as a "name space ID" for all UUIDs generated from names in that name space; see Appendix C for some pre-defined values.

File: lib/puppet/parser/functions/uuid_generate.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    hash_input = ns_uuid + fqdn
+    unless seed.nil?
+      hash_input += "\x00" + seed
```
Comment (hunner) [inline lib/puppet/parser/functions/uuid_generate.rb]: If the seed can be any string, why prefix with `'\x00'`?

File: lib/puppet/parser/functions/uuid_generate.rb (hunk 2 above)
Comment (petems) [inline lib/puppet/parser/functions/uuid_generate.rb]: Yeah, this was a holdover from the older FQDN specific code

File: lib/puppet/parser/functions/uuid_generate.rb (hunk 4 above)
Comment (petems) [inline lib/puppet/parser/functions/uuid_generate.rb]: TBH I'm not sure about this bit, I was leveraging an existing bit of code

Comment (petems) [inline lib/puppet/parser/functions/uuid_generate.rb]: What would need to change here?

**Pull request merged**